source/event_stream_rpc_client_connection.c

```diff
@@ -0,0 +1,292 @@
+}
+
+static void s_connection_destroy_if_ready(struct connection_binding *connection) {
+    bool destroy;
+
+    if (connection->native) {
+        if (connection->capsule_destroyed) {
+            if (connection->shutdown_complete) {
+                /* python class has been GC'd and native connection has finished shutdown */
+                destroy = true;
+            } else {
+                /* python class has been GC'd but native connection still neeeds to shut down */
+                aws_event_stream_rpc_client_connection_close(connection->native, AWS_ERROR_SUCCESS);
+                destroy = false;
+            }
+        } else {
+            /* native connection has shut down, but python class still exists */
+            destroy = false;
+        }
+    } else {
+        /* native connection setup failed */
+        destroy = true;
+    }
+
+    if (destroy) {
+        Py_XDECREF(connection->on_setup);
+        Py_XDECREF(connection->on_shutdown);
+        Py_XDECREF(connection->on_protocol_message);
+        aws_event_stream_rpc_client_connection_release(connection->native);
+        aws_mem_release(aws_py_get_allocator(), connection);
+    }
+}
+
+static void s_on_protocol_message(
+    struct aws_event_stream_rpc_client_connection *native,
+    const struct aws_event_stream_rpc_message_args *message_args,
+    void *user_data) {
+
+    (void)native;
+    struct connection_binding *connection = user_data;
+
+    PyGILState_STATE state;
+    if (aws_py_gilstate_ensure(&state)) {
+        return; /* Python has shut down. Nothing matters anymore, but don't crash */
+    }
+
+    /* TODO: create Python headers from C headers */
+    PyObject *headers_py = Py_None;
+    Py_INCREF(headers_py);
+
+    PyObject *result = PyObject_CallFunction(
+        connection->on_protocol_message,
+        "(Oy#iI)",
+        headers_py,
+        message_args->payload->buffer,
+        message_args->payload->len,
+        message_args->message_type,
+        message_args->message_flags);
+    if (result) {
+        Py_DECREF(result);
+    } else {
+        /* Callback might fail during application shutdown */
+        PyErr_WriteUnraisable(PyErr_Occurred());
+    }
+
+    Py_DECREF(headers_py);
+
+    PyGILState_Release(state);
+}
+
+PyObject *aws_py_event_stream_rpc_client_connection_close(PyObject *self, PyObject *args) {
+    (void)self;
+    PyObject *capsule;
+    if (!PyArg_ParseTuple(args, "O", &capsule)) {
+        return NULL;
+    }
+
+    struct connection_binding *connection = PyCapsule_GetPointer(capsule, s_capsule_name);
+    if (!connection) {
+        return NULL;
+    }
+
+    aws_event_stream_rpc_client_connection_close(connection->native, AWS_OP_SUCCESS);
+    Py_RETURN_NONE;
+}
+
+PyObject *aws_py_event_stream_rpc_client_connection_is_open(PyObject *self, PyObject *args) {
+    (void)self;
+    PyObject *capsule;
+    if (!PyArg_ParseTuple(args, "O", &capsule)) {
+        return NULL;
```

Comment:
Should be returning false?

Reply:
For C functions bound to python, error handling works like this:
If there's an error, set the exception and return `NULL`.
In this case, PyArg_ParseTuple() has already set an exception about failure to parse the arguments passed into this fucntion.